File: p012.py
#!/usr/bin/python

import functools


def factors(n):
    return set(functools.reduce(list.__add__, ([i, n//i] for i in range(1, int(n**0.5) + 1) if n % i == 0)))

num = 1

while True:
    prod = sum(range(1, num + 1))
    t = factors(prod)
    if len(t) >= 500:
        break
    num += 1

print(prod)

File: p012.rs

fn num_divisors(n: usize) -> usize {
    let mut cnt = 0;
    for i in (1..n + 1) {
        if n % i == 0 {
            cnt += 1;
        }
    }
    cnt
}

fn main() {

    let mut num = 1;
    let mut prod: usize;

    loop {
        prod = (1..num + 1).fold(0, |x, sum| sum + x);
        let t = num_divisors(prod);
        if t >= 500 {
            break;
        }
        println!("{} {}", prod, t);
        num += 1
    }

    println!("{}", prod);
}
